src/main/java/com/zaxxer/hikari/util/ConcurrentBag.java

```diff
@@ -245,13 +245,14 @@ public void close()
     */
    public List<T> values(final int state)
    {
-      final ArrayList<T> list = new ArrayList<>(sharedList.size());
-      for (final T reference : sharedList) {
-         if (reference.getState() == state) {
-            list.add(reference);
+      final int size = sharedList.size();
+      final ArrayList<T> list = new ArrayList<>(size);
+      for (int i = 0; i < size; i++) {
+         T entry = sharedList.get(i);
```

Comment:
for-each throws ConcurrentModificationException see  item 7 here https://today.java.net/pub/a/today/2006/11/07/nuances-of-java-5-for-each-loop.html